src/storage/src/source/persist_source.rs

```diff
@@ -58,140 +58,165 @@ where
-                            for cap in cap_set.iter() {
-                                // NOTE: The nightly `drain_filter()` would use
-                                // less allocations than this. Should switch to
-                                // it once it's available in stable rust.
-                                let (mut to_emit, remaining_updates) =
-                                    updates.into_iter().partition(|(_update, ts, _diff)| {
-                                        PartialOrder::less_equal(cap.time(), ts)
-                                    });
-                                updates = remaining_updates;
-
-                                let mut session = output.session(&cap);
-                                session.give_vec(&mut to_emit);
-                            }
-
-                            assert!(
-                                updates.is_empty(),
-                                "did not have matching Capability for updates: {:?}",
-                                updates
-                            );
-                        }
                         Some(Err::<_, ExternalError>(e)) => {
-                            // TODO(petrosagg): error handling
-                            panic!("unexpected error from persist {e}");
+                            panic!("unexpected error from persist {e}")
                         }
+                        // We never expect any further output from
+                        // `pinned_stream`, so propagate that information
+                        // downstream.
                         None => {
-                            // Empty out the `CapabilitySet` to indicate that we're done.
                             cap_set.downgrade(&[]);
                             return;
                         }
                     }
-                    if timer.elapsed() > YIELD_INTERVAL {
-                        activator.activate();
-                        break;
-                    }
                 }
             }
         });
 
-    let (ok_stream, err_stream) = timely_stream.ok_err(|x| match x {
+    let mut builder = OperatorBuilder::new(
+        format!(
+            "persist_source: sharded reader {} of {:?}",
+            worker_index, source_id
+        ),
+        scope.clone(),
+    );
+    let operator_info = builder.operator_info();
+    let dist = |&(i, _): &(usize, ReaderEnrichedHollowBatch<Timestamp>)| Timestamp::cast_from(i);
+    let mut input = builder.new_input(&inner, Exchange::new(dist));
+    let (mut output, output_stream) = builder.new_output();
+
+    // Bound execution of operator to prevent a single operator from hogging
+    // the CPU if there are many messages to process
+    let activator = scope.activator_for(&operator_info.address[..]);
+
+    builder.build_async(
+        scope.clone(),
+        async_op!(|initial_capabilities, _frontiers| {
+            let mut read = persist_clients
```

Comment:
That is what [this code](https://github.com/MaterializeInc/materialize/pull/13849/files#diff-6f304b700fb56de26b66e09ff27e8c3dcea8bfff6aea64da3c291a4a6b24b369R616-R618) in `fetch_batch` attempts to do. Something I missed here?

Reply:
ah! that feels to me like a confusing contract for fetch_batch. (partial evidence: it's surprising enough to me that I didn't even realize it was happening while looking over this PR)

Reply:
Is it just the name of the function that you drew assumptions from? Would something like `process_batch` be more illustrative from your perspective?

Reply:
no, I think it's just an odd contract to have for a method that exists on ReadHandle, which is pretty general purpose. 
I let this mull for the last hour or so and I think everything clears up if we have a new something (BatchStreamFetcher?) that wraps the ReadHandle and this `fetch_batch` method is only available on that (think the pattern that Listen has now)

I actually think I'd like to introduce something exactly along those lines so that we can abstract over pipelining batch fetches, etc. that should 100% be a followup to this PR, so, in conclusion, I think what you have now is fine to merge as-is. do me a favor a leave a TODO next to that code you linked above about it potentially being surprising for a ReadHandle user, please

@aljoscha one thing I realized mulling all this over is that our story around the seqno capabilities is a little underdeveloped here. the first operator is handing our references to things, how do we ensure the second operator doesn't have those things gc'd out from under it?

Reply:
I think we might not currently be ensuring that! The immediate solution that comes to mind is that persist does ref-counting for capabilities, similar to how timely does. The first stage operator (the one that listens for changes) would records that it hands out a capability for a given seqno, the reader operator then records that it is done with it, reducing the ref count. 

This is just a strawman, for now!

Reply:
More spitballing: The "subscribe" operator is the only one that really cares about holding back the `since` frontier. The "fetcher" operators don't care about the `since` frontier at all but they _do_ care about holding back the `SeqNo` frontier.

What we could do is: the fetcher operators turn their `ReadHandle` into a special `FetchHandle`, which immediately let's the `since` go to `[]`. We track the frontier of the required `SeqNo` from the subscribe operator to the fetch operator and the fetcher operators _explicitly_ downgrade the `SeqNo` frontier using the `FetchHandle`.

In state, the `FetchHandle` could look like a regular `ReadHandle` that just has it's `since` at `[]` or we could track them separately in state. I'm sure we all don't want the added complexity of the latter, though.

Reply:
How would the ref count work? We can't use an Arc because it has to be ExchangeData. Are you thinking of tracking it state? That would be pretty expensive and unfortunate.

I like something like the FetchHandle, I've been imagining something vaguely like it to abstract over pipelining the batch fetches, anyway. 100% not suggesting adding FetchHandle as part of a scope increase for sean in this PR. But, given that things are working right now in main, I'm a little hesitant to merge and potentially create another fire for persist without 1. being comfortable with the design for how we'll fix this and 2. having someone lined up to do it as a fast-follow. I also don't want to leave this PR open to rot if we don't have someone right away for (2).

@aljoscha can we sync up on what we want to do here as soon as I'm on tomorrow?

Reply:
fwiw this is my top priority and if there is follow-on work to make this Good™, I am glad to do it. I don't think adding a `FetchHandle` is much work and I'm glad to make it part of this PR, or an immediate follow up.

Reply:
:thumbsup: I think I do need to have the design discussion with aljoscha before we can make any concrete decisions on next steps